zerver/models/push_notifications.py

```diff
@@ -40,3 +47,68 @@ class PushDeviceToken(AbstractPushDeviceToken):
 
     class Meta:
         unique_together = ("user", "kind", "token")
+
+
+class AbstractPushDevice(models.Model):
+    class TokenKind(models.TextChoices):
+        APNS = "apns", "APNs"
+        FCM = "fcm", "FCM"
+
+    token_kind = models.CharField(max_length=4, choices=TokenKind.choices)
+
+    # 64-bit random integer ID generated by the client; will only be
+    # guaranteed to be unique within the client's own table of accounts.
+    push_account_id = models.BigIntegerField()
+
+    class Meta:
+        abstract = True
+
+
+class PushDevice(AbstractPushDevice):
+    """Core zulip server table storing registrations that are potentially
+    registered with the mobile push notifications bouncer service.
+
+    Each row corresponds to an account on an install of the app
+    that has attempted to register with the bouncer to receive
+    mobile push notifications.
+    """
+
+    # The user on this server to whom this PushDevice belongs.
+    user = models.ForeignKey(UserProfile, on_delete=CASCADE)
+
+    # Key to use to encrypt notifications for delivery to this device.
+    push_public_key = models.TextField()
+
+    # ID to identify the corresponding record on RemotePushDevice.
+    # Set to NULL while registration to bouncer is in progress.
+    bouncer_device_id = models.BigIntegerField(null=True)
+
+    class ErrorCode(models.TextChoices):
+        INVALID_BOUNCER_PUBLIC_KEY = InvalidBouncerPublicKeyError.code.name
+        INVALID_ENCRYPTED_PUSH_REGISTRATION_ERROR = InvalidEncryptedPushRegistrationError.code.name
+        REQUEST_EXPIRED_ERROR = RequestExpiredError.code.name
+
+    # The error code returned when registration to bouncer fails.
+    # Set to NULL if the registration is in progress or successful.
+    error_code = models.CharField(max_length=30, choices=ErrorCode.choices, null=True)
```

Comment:
Let's make this max_length 100. It will be very annoying if it's too short, since we can update the mobile apps and the bouncer much more easily than all self-hosted servers. Since it'll be a `VARCHAR`, there's no real benefit for setting this low, but I guess maybe some small anti-DoS benefit for it being finite.